### DISCUSSION TITLE
Does Executorch support on-device training/learning

### DISCUSSION THREAD

**Discussion by knn1989** [pytorch/executorch#8221] [Q&A] [answered]
Does Executorch support on-device training (online learning/model update on edge devices)? If yes, how to enable this? Thanks.

**Comment by guangy10** [comment]
This is one area the team is exploring. @JacobSzwejbka could comment more

**Comment by knn1989** [comment]
It would be greatly appreciated if this feature could be supported. I have attempted numerous times to transition from using TensorFlow Lite to PyTorch for edge device deployment, but the absence of features like this has hindered my efforts. Thank you.

**Comment by jayzhan211** [comment]
I'm quite curious, usually training unlike inferencing expect much more computing, why does on-device training helpful? Is it because of that reduces the develop iteration cycle for training smaller model?

**Comment by JacobSzwejbka** [comment]
Its under active development, and we should have something to share soon. From your side @knn1989 can you tell me a little bit about what your use case looks like?

**Comment by knn1989** [comment]
@JacobSzwejbka, thank you for your response. The on-device learning capability offers numerous advantages, such as enhanced personalization, increased security, and the ability to process data offline.

My use cases involve training or fine-tuning small models, using different techniques ranging from reinforcement learning, supervised learning to federated learning, where rewards or labels are accessible to improve and update the model. Edge devices such as smartphones, equipped with GPUs and NPUs, are well-suited for this purpose. While TensorFlow Lite supports these features, PyTorch has yet to incorporate similar capabilities.

**Comment by JacobSzwejbka** [comment, marked as answer]
https://github.com/pytorch/executorch/blob/main/extension/training/README.md 

We have added training support here. Please give it a shot and let us know what sort of improvements or further development you are interested in!
